authorization.go

```diff
@@ -0,0 +1,204 @@
+	CNs []string
+}
+
+func (allow AllowSpecificOUandCNs) IsAuthorized(
+	clientOU []string, clientCN string,
+) error {
+	if len(allow.OUs) > 0 {
+		if err := allowedOU(allow.OUs, clientOU); err != nil {
+			return err
+		}
+	}
+	if len(allow.CNs) > 0 {
+		if err := allowedCN(allow.CNs, clientCN); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (allow AllowSpecificOUandCNs) IsAuthorizedForSite(
+	clientOU []string, clientCN, uriSite string,
+) error {
+	// Site resources are not handled separately. Fall back to the behavior
+	// of IsAuthorized
+	return allow.IsAuthorized(clientOU, clientCN)
+}
+
+// RestrictSiteOUs is an instance of AuthorizationCheck that restricts
+// site resource access when the client is in certain OUs.
+// The OUs which this applies to are provided by AppliedOUs.
+// If a client within on of the AppliedOUs requests access to a site resource
+// (i.e. a resource with the `site` URI Parameter), their CN is parsed to obtain their
+// site ID. This site ID must be an exact match of the site ID they are requesting.
+// If the site ID determined from the client's CN does not match the requested resource
+// site ID, the request is denied.
+// If the client's OUs do not match any of `AppliedOUs` the request is allowed.
+// This is intended to be used along with AllowSpecificOUandCNs:
+//    Use AllowSpecificOUandCNs to restrict access to a broad group of OUs
+//    Use RestrictSiteAccess to lock specific OUs into only access their own sites.
+// For example:
+//
+//   opts := certauth.Opts{
+//       AuthorizationChecks: []AuthorizationCheck{
+//           AllowOUs([]string{"admin", "site", "employee"}),
+//           RestrictSiteAccess{
+//               AppliedOUs: []string{"site"},
+//       },
+//   }
+//
+// This example would require *all* requests to present an OU of `admin`, `site`, or
+// `employee` and additionally, requests with an OU of `site` would be prevented from
+// accessing site-specific resources that were not part of their own site.
+type RestrictSiteAccess struct {
+	AppliedOUs []string
+
+	// NOTE(zeal): maybe add a field here to allow the server to choose how the
+	//             site is parsed from the CN?
+}
+
+func (check RestrictSiteAccess) IsAuthorized(clientOU []string, clientCN string) error {
+	// This check only operates on site resources.
+	return nil
+}
+
+func (check RestrictSiteAccess) IsAuthorizedForSite(
+	clientOU []string, clientCN, uriSite string,
+) error {
+	site, _, err := ParseSiteEnvFromCN(clientCN)
+	if err != nil {
+		return err
+	}
+	if site == uriSite {
+		return nil
+	}
+	return fmt.Errorf("Site %q is not authorized to site %q resources", site, uriSite)
+}
+
+// ParseSiteEnvFromCN parses a site id and environment from the provided CN.
+// Also validates that the site ID is a valid UUID.
+// Returns (site, environment, nil) if the clientCN is valid.
+// Returns ("", "", err) if an error occurs.
+func ParseSiteEnvFromCN(clientCN string) (string, string, error) {
```

Comment:
I would make named return instead of docstring:
```suggestion
func ParseSiteEnvFromCN(clientCN string) (site string, environment string, err error) {
```
https://golang.org/doc/effective_go#named-results

Or return some struct
```
type ParseSiteEnvFromCNResp struct {
    Site string
    Environment string
}
```

Reply:
-1 on named return. Feature shouldn't exist if less really is exponentially more, as language designers claim.
+1 on struct. Maybe just one func that takes the whole subject and returns a struct holding all the subject info AND the stuff parsed out.

Reply:
> I would make named return instead of docstring:

Lol. I originally wrote it this way using a named return cause it's much clearer what each thing is returning.

But on previous PRs, several people got angry at me for using named returns like this, so I restructured this to avoid comments like that.

I'd much rather use a named return here as I think it's clearer what is returned by the function, and the function is simple enough I'm not too worried about the issues around named returns.

But honestly, the way it is now, it's obvious what's returned by the comment on line 161 and 162. And by the return statements throughout the function.

Wrapping the return in the struct might be helpful if more information gets embedded in the CN later that we need to parse out, but I don't think it makes this function more readable as-is.

Reply:
Sorry, I'm a newbie here, just wanted to mention [the common recommendation](https://golang.org/doc/effective_go#named-results).

And yes, named returned is bad in case of returning two values of different types.

Reply:
By the way, our [Pantheon Go Style Guide](https://github.com/pantheon-systems/go-style-guide/blob/master/style.md) is referencing [Effective Go](https://golang.org/doc/effective_go.html)

Reply:
OK. So no named return, just a question about struct.

Reply:
I'm gonna leave this as-is. If you strongly disagree please un-resolve and we can discuss.